pkg/controller/jobcontrol.go

```diff
@@ -0,0 +1,271 @@
+		return nil, false, c.deleteOldJobs(ownerKey, oldJobs, logger)
+	}
+
+	if currentJob != nil {
+		return currentJob, false, nil
+	}
+
+	if jobFactory != nil {
+		job, err := c.createJob(ownerKey, owner, jobFactory, logger)
+		return job, job != nil, err
+	}
+
+	return nil, false, nil
+}
+
+func (c *jobControl) ObserveOwnerDeletion(ownerKey string) {
+	c.expectations.DeleteExpectations(ownerKey)
+}
+
+func (c *jobControl) ObserveJobCreation(ownerKey string, job *kbatch.Job) {
+	c.expectations.CreationObserved(ownerKey)
+}
+
+func (c *jobControl) ObserveJobDeletion(ownerKey string, job *kbatch.Job) {
+	c.expectations.DeletionObserved(ownerKey, jobKey(job))
+}
+
+func (c *jobControl) IsControlledJob(job *kbatch.Job) bool {
+	return strings.HasPrefix(job.Name, c.jobPrefix)
+}
+
+func (c *jobControl) createJob(ownerKey string, owner metav1.Object, jobFactory JobFactory, logger log.FieldLogger) (*kbatch.Job, error) {
+	logger.Infof("Creating new %q job", c.jobPrefix)
+
+	name := names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s%s-", c.jobPrefix, owner.GetName()))
+
+	job, configMap, err := jobFactory.BuildJob(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if job == nil {
+		return nil, nil
+	}
+
+	ownerRef := metav1.NewControllerRef(owner, c.ownerKind)
+
+	job.OwnerReferences = append(job.OwnerReferences, *ownerRef)
+	if job.Annotations == nil {
+		job.Annotations = map[string]string{}
+	}
+	job.Annotations[clusteroperator.OwnerGenerationAnnotation] = fmt.Sprintf("%d", owner.GetGeneration())
+
+	cleanUpConfigMap := true
+	if configMap != nil {
+		configMap.OwnerReferences = append(configMap.OwnerReferences, *ownerRef)
+
+		configMap, err = c.kubeClient.CoreV1().ConfigMaps(owner.GetNamespace()).Create(configMap)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if configMap != nil && cleanUpConfigMap {
+				if err := c.kubeClient.CoreV1().ConfigMaps(configMap.Namespace).Delete(configMap.Name, &metav1.DeleteOptions{}); err != nil {
+					logger.Warnf("Could not delete config map %v/%v", configMap.Namespace, configMap.Name)
+				}
+			}
+		}()
+	}
+
+	if err := c.expectations.ExpectCreations(ownerKey, 1); err != nil {
+		return nil, err
+	}
+
+	newJob, err := c.kubeClient.BatchV1().Jobs(owner.GetNamespace()).Create(job)
+	if err != nil {
+		// If an error occurred creating the job, remove expectation
+		c.expectations.CreationObserved(ownerKey)
+		return nil, err
+	}
+
+	cleanUpConfigMap = false
+
+	return newJob, nil
+}
+
+func (c *jobControl) deleteOldJobs(ownerKey string, oldJobs []*kbatch.Job, logger log.FieldLogger) error {
+	keysToDelete := []string{}
+	for _, job := range oldJobs {
+		keysToDelete = append(keysToDelete, jobKey(job))
+	}
+	c.expectations.ExpectDeletions(ownerKey, keysToDelete)
+
+	logger.Infof("Deleting old jobs: %v", keysToDelete)
+
+	var errCh chan error
+
+	var wg sync.WaitGroup
+	wg.Add(len(oldJobs))
+	for i, ng := range oldJobs {
+		go func(ix int, job *kbatch.Job) {
+			defer wg.Done()
```

Comment:
if there's an error, it doesn't look like we'd ever call this

Reply:
I'm not following. Doesn't the defer *always* get called, error or not?

Reply:
Is your issue that the `errCh` is not being created? That is certainly a problem that I should resolve. It is also a problem in `clusterController`.

Reply:
the problem is that sending to errCh will block until you read from it, but you won't get to the part that reads from it until all goroutines have called wg.Done(), essentially deadlocking

Reply:
Sending to `errCh` won't block unless `errCh` is already full. The size of `errCh` is now set to the length of `oldJobs`, so the sender will never block.

Reply:
ahh buffered channel :) 👍

Reply:
I immediately assumed it was unbuffered, but I guess you were missing the make call